pkg/network/proxy/proxy.go

```diff
@@ -221,73 +236,65 @@ func (proxy *OsdnProxy) updateEgressNetworkPolicy(policy networkv1.EgressNetwork
 
 	// Add/Update/Delete firewall rules for the namespace
 	if len(firewall) > 0 {
-		if _, ok := proxy.firewall[ns]; !ok {
-			item := &proxyFirewallItem{}
-			item.namespaceFirewalls = make(map[ktypes.UID][]firewallItem)
-			item.activePolicy = nil
-			proxy.firewall[ns] = item
-		}
-		proxy.firewall[ns].namespaceFirewalls[policy.UID] = firewall
-	} else if _, ok := proxy.firewall[ns]; ok {
-		delete(proxy.firewall[ns].namespaceFirewalls, policy.UID)
-		if len(proxy.firewall[ns].namespaceFirewalls) == 0 {
-			delete(proxy.firewall, ns)
-		}
+		ns.firewalls[policy.UID] = firewall
+	} else {
+		delete(ns.firewalls, policy.UID)
 	}
 
 	// Set active policy for the namespace
-	if ref, ok := proxy.firewall[ns]; ok {
-		if len(ref.namespaceFirewalls) == 1 {
-			for uid := range ref.namespaceFirewalls {
-				ref.activePolicy = &uid
-				klog.Infof("Applied firewall egress network policy: %q to namespace: %q", uid, ns)
-			}
-		} else {
-			ref.activePolicy = nil
+	if len(ns.firewalls) == 1 {
+		for uid := range ns.firewalls {
+			ns.activePolicy = &uid
+			klog.Infof("Applied firewall egress network policy: %q to namespace: %q", uid, policy.Namespace)
+		}
+	} else {
+		ns.activePolicy = nil
+
+		if len(ns.firewalls) > 1 {
 			// We only allow one policy per namespace otherwise it's hard to determine which policy to apply first
-			utilruntime.HandleError(fmt.Errorf("Found multiple egress policies, dropping all firewall rules for namespace: %q", ns))
+			utilruntime.HandleError(fmt.Errorf("Found multiple egress policies, dropping all firewall rules for namespace: %q", policy.Namespace))
 		}
 	}
 
 	// Update endpoints
-	for _, pep := range proxy.allEndpoints {
-		if pep.endpoints.Namespace != policy.Namespace {
-			continue
-		}
-
+	for _, pep := range ns.endpoints {
 		wasBlocked := pep.blocked
-		pep.blocked = proxy.endpointsBlocked(pep.endpoints)
+		pep.blocked = proxy.endpointsBlocked(ns, pep.endpoints)
 		switch {
 		case wasBlocked && !pep.blocked:
 			proxy.baseProxy.OnEndpointsAdd(pep.endpoints)
 		case !wasBlocked && pep.blocked:
 			proxy.baseProxy.OnEndpointsDelete(pep.endpoints)
 		}
 	}
+
+	if len(ns.firewalls) == 0 {
+		proxy.maybeDeleteNamespace(policy.Namespace, ns)
+	}
 }
 
-func (proxy *OsdnProxy) firewallBlocksIP(namespace string, ip net.IP) bool {
-	if ref, ok := proxy.firewall[namespace]; ok {
-		if ref.activePolicy == nil {
-			// Block all connections if active policy is not set
-			return true
-		}
+func (ns *proxyNamespace) firewallBlocksIP(ip net.IP) bool {
```

Comment:
this is not thread safe

Reply:
nevermind, it is called from EndpointsAdd and EndpointsUpdated that hold the lock

Reply:
I should be clearer about what functions assume the lock is already held...